Pr branch

TasnimAwashreh opened the pull request.

TasnimAwashreh committed:
- Added git ignore

TasnimAwashreh committed:
- Implemented Observer Pattern for the subject and bots

TasnimAwashreh committed:
- Added Configuration
- Moved config file to the project root
- Fixed the structure of the project
- Added a constants file for strings
- added an input parser where the user can enter multiple lines
- Added a format reader that chooses between xml or json depending on the user's input

TasnimAwashreh committed:
- New branch for PR

RandaZraik commented on WeatherMonitoringAndReportingService/config.json:
```diff
@@ -0,0 +1,17 @@
+{
```
this file we call it `appsettings.json` usually

RandaZraik commented on WeatherMonitoringAndReportingService/Logic/Services/BotLoader.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public void LoadBots(ISubject subject)
+        {
+            var rainBot = _config.GetSection("RainBot").Get<RainBot>();
```
usually we don't use our models/dtos to load configs, we should create different classes for configs.

RandaZraik commented on WeatherMonitoringAndReportingService/Logic/Subjects/Subject.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace WeatherMonitoringAndReportingService.Logic.Subjects
+{
+    public class Subject : ISubject
```
A better name is `WeatherStation`

RandaZraik commented on WeatherMonitoringAndReportingService/Logic/Readers/JsonFormatReader.cs:
```diff
@@ -0,0 +1,13 @@
+    {
+        public WeatherData? ReadWeatherData(string userInput)
+        {
+            return JsonSerializer.Deserialize<WeatherData>(userInput);
```
deserialization can throws invalid format exceptions, lets try-catch and return null if sth is off

RandaZraik commented on WeatherMonitoringAndReportingService/Logic/Readers/XMLFormatReader.cs:
```diff
@@ -0,0 +1,16 @@
+    {
+        public WeatherData? ReadWeatherData(string userInput)
+        {
+            var serializer = new XmlSerializer(typeof(WeatherData));
```
same comment as above

RandaZraik commented on WeatherMonitoringAndReportingService/App/Program.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        using var scope = serviceProvider.CreateScope();
+
+        var BotSystem = scope.ServiceProvider.GetRequiredService<BotSystem>();
```
camelCase

RandaZraik commented on WeatherMonitoringAndReportingService/Data/Models/WeatherData.cs:
```diff
@@ -0,0 +1,19 @@
+﻿namespace WeatherMonitoringAndReportingService.Data.Models
+{
+    public class WeatherData
```
Data folder is really just for storage/db. Here we don't really need that layer. This model belongs to the logic layer

RandaZraik commented on WeatherMonitoringAndReportingService/Logic/Readers/ChooseFormatReader.cs:
```diff
@@ -0,0 +1,14 @@
+﻿namespace WeatherMonitoringAndReportingService.Logic.Readers
+{
+    public static class ChooseFormatReader
```
for handling the readers we should be using the strategy pattern as specified in the exercise

RandaZraik commented on WeatherMonitoringAndReportingService/Logic/Observers/Bots/SnowBot.cs:
```diff
@@ -0,0 +1,15 @@
+
+        public override void Update(WeatherData weatherData)
+        {
+            if (weatherData.Temperature < TemperatureThreshold)
```
check if enabled as well, same for other bots

TasnimAwashreh committed:
- Renamed Subject to WeatherStation
- changed config.json to appsettings.json
- Moved WeatherData to Domain
- Bots now check if theyre enabled to update

TasnimAwashreh committed:
- Added DTOs

TasnimAwashreh committed:
- Surrounded readers with try catch

TasnimAwashreh committed:
- Updated Reader classes
